server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/lib/AbstractStateCompositeProcessor.java

```diff
@@ -66,42 +66,50 @@ public void removeListener(CompositeProcessorListener listener) {
     }
 
     @Override
-    public void configure(HierarchicalConfiguration config) throws ConfigurationException {
+    public void configure(HierarchicalConfiguration config) {
         this.config = config;
         this.enableJmx = config.getBoolean("[@enableJmx]", true);
 
     }
 
     @Override
     public void service(Mail mail) throws MessagingException {
-        long start = System.currentTimeMillis();
-        MessagingException ex = null;
         MailProcessor processor = getProcessor(mail.getState());
 
         if (processor != null) {
-            LOGGER.debug("Call MailProcessor {}", mail.getState());
-            try {
-                processor.service(mail);
-
-                if (Mail.GHOST.equals(mail.getState())) {
-                    LifecycleUtil.dispose(mail);
-                }
-                /*
-                 * // check the mail needs further processing if
-                 * (Mail.GHOST.equalsIgnoreCase(mail.getState()) == false) {
-                 * service(mail); } else { LifecycleUtil.dispose(mail); }
-                 */
-            } catch (MessagingException e) {
-                ex = e;
-                throw e;
-            } finally {
-                long end = System.currentTimeMillis() - start;
-                for (CompositeProcessorListener listener : listeners) {
-                    listener.afterProcessor(processor, mail.getName(), end, ex);
-                }
-            }
+            doService(mail, processor);
         } else {
-            throw new MessagingException("No processor found for mail " + mail.getName() + " with state " + mail.getState());
+            MailProcessor errorProcessor = getProcessor(Mail.ERROR);
```

Comment:
this function does too many things and takes responsibility of getProcessor.
Could you move this "fallback" code to getProcessor? That way, we would always call doService.